### PR TITLE
Fix flaky keepalive ping test

### LINKED ISSUE #24797: Test failure: PING_NoKeepAliveTimeout_DoesNotResetKeepAliveTimeout
https://dev.azure.com/dnceng/public/_build/results?buildId=764711&view=ms.vss-test-web.build-test-results-tab&runId=23761986&paneView=debug&resultId=106734
https://dev.azure.com/dnceng/public/_build/results?buildId=765006&view=ms.vss-test-web.build-test-results-tab&runId=23773364&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab&resultId=121827

```
Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2ConnectionErrorException : HTTP/2 connection error (FRAME_SIZE_ERROR): The received frame size of 36 exceeds the limit 8.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2FrameReader.TryReadFrame(ReadOnlySequence`1& buffer, Http2Frame frame, UInt32 maxFrameSize, ReadOnlySequence`1& framePayload) in /_/src/Shared/ServerInfrastructure/Http2/Http2FrameReader.cs:line 48
   at Microsoft.AspNetCore.Server.Kestrel.Core.Tests.Http2TestBase.ReceiveFrameAsync(UInt32 maxFrameSize) in /_/src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TestBase.cs:line 1108
   at Microsoft.AspNetCore.Server.Kestrel.Core.Tests.Http2TestBase.ExpectAsync(Http2FrameType type, Int32 withLength, Byte withFlags, Int32 withStreamId) in /_/src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TestBase.cs:line 1134
   at Microsoft.AspNetCore.Server.Kestrel.Core.Tests.Http2TimeoutTests.PING_NoKeepAliveTimeout_DoesNotResetKeepAliveTimeout() in /_/src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs:line 172
--- End of stack trace from previous location ---
```
```
[0.000s] [TestLifetime] [Information] Starting test PING_NoKeepAliveTimeout_DoesNotResetKeepAliveTimeout at 2020-08-11T13:03:32
[0.016s] [Microsoft.AspNetCore.Server.Kestrel] [Verbose] Connection id "null" sending SETTINGS frame for stream ID 0 with length 18 and flags NONE
[0.016s] [Microsoft.AspNetCore.Server.Kestrel] [Verbose] Connection id "null" sending WINDOW_UPDATE frame for stream ID 0 with length 4 and flags 0x0
[0.016s] [Microsoft.AspNetCore.Server.Kestrel] [Verbose] Connection id "null" received SETTINGS frame for stream ID 0 with length 0 and flags NONE
[0.016s] [Microsoft.AspNetCore.Server.Kestrel] [Verbose] Connection id "null" sending SETTINGS frame for stream ID 0 with length 0 and flags ACK
[0.016s] [Microsoft.AspNetCore.Server.Kestrel] [Verbose] Connection id "null" received HEADERS frame for stream ID 1 with length 318 and flags END_HEADERS
[0.016s] [Microsoft.AspNetCore.Server.Kestrel] [Verbose] Connection id "null" received PING frame for stream ID 0 with length 8 and flags NONE
[0.016s] [Microsoft.AspNetCore.Server.Kestrel] [Verbose] Connection id "null" sending HEADERS frame for stream ID 1 with length 36 and flags END_STREAM, END_HEADERS
[0.016s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id "null", Request id ":00000001": started reading request body.
[0.016s] [Microsoft.AspNetCore.Server.Kestrel] [Verbose] Connection id "null" sending PING frame for stream ID 0 with length 8 and flags ACK
[0.016s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id "null", Request id ":00000001": done reading request body.
[0.016s] [Microsoft.AspNetCore.Server.Kestrel] [Information] Connection id "null", Request id ":00000001": the application completed without reading the entire request body.
[0.016s] [Microsoft.AspNetCore.Server.Kestrel] [Verbose] Connection id "null" sending RST_STREAM frame for stream ID 1 with length 4 and flags 0x0
[0.016s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id "null" is closed. The last processed stream ID was 1.
[0.016s] [Microsoft.AspNetCore.Server.Kestrel] [Verbose] Connection id "null" sending GOAWAY frame for stream ID 0 with length 8 and flags 0x0
[0.047s] [TestLifetime] [Information] Finished test PING_NoKeepAliveTimeout_DoesNotResetKeepAliveTimeout in 0.033467s
```
Looks like there's a race in the test where the response headers are sent before the ping ack.

@JamesNK

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes #24797

### CODE REVIEW

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs [hunk 1]
```diff
@@ -156,13 +157,14 @@ public async Task PING_NoKeepAliveTimeout_DoesNotResetKeepAliveTimeout()
...
 
-            await InitializeConnectionAsync(_noopApplication);
+            await InitializeConnectionAsync(_echoApplication);
 
             // Connection starts and sets keep alive timeout
             _mockTimeoutControl.Verify(c => c.SetTimeout(It.IsAny<long>(), TimeoutReason.KeepAlive), Times.Once);
             _mockTimeoutControl.Verify(c => c.ResetTimeout(It.IsAny<long>(), TimeoutReason.KeepAlive), Times.Never);
             _mockTimeoutControl.Verify(c => c.CancelTimeout(), Times.Never);
 
+            // Stream will stay open because it is waiting for request body to end
```
Comment (Tratcher) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs:167]: Do you need to send that body eventually to unblock the app for cleanup? Or does server tear down kill the body and let it exit?

**Review by Tratcher — APPROVED** [review]

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs (hunk 1 above)
Comment (JamesNK) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs:167]: I've always assumed that `Http2TestBase.Dispose()` cleans up any outstanding requests.

Comment (Tratcher) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs:167]: Can you put a break point in the echo app to confirm that?

Comment (JamesNK) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs:167]: Hmm, the connection appears to have ended but it appeared the echo app never progressed. I'll put an explicit end stream in the test.

**Pull request merged**